src/main/java/gndata/lib/util/ChangeHelper.java

```diff
@@ -12,46 +13,195 @@
-        Change ch = new Insert(newObject);
-        applyChange(ch);
+    public Change get(int index) {
+        return changes.get(index);
     }
 
-    public void delete(String uri) {
-        Change ch = new Delete(uri);
-        applyChange(ch);
+    public int size() {
+        return changes.size();
     }
 
-    // TODO add more methods
+    /*   Private helper functions   */
+
+    private static boolean isUpdate(Change removed, Change added) {
+
+        // assume update is always first remove and then add
+
+        if (added.getChange().size() != removed.getChange().size()) {
+            return false;
+        }
+
+        StmtIterator iter = removed.getChange().listStatements();
+        while (iter.hasNext()) {
+            Statement st = iter.nextStatement();
+
+            if (added.getChange()
+                    .listStatements(st.getSubject(), st.getPredicate(), (RDFNode) null)
+                    .toList().size() != 1 ) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean isLastUpdate() {
+        return position > 0 && isUpdate(changes.get(position - 1), changes.get(position));
+    }
+
+    private boolean isNextUpdate() {
+        boolean posCheck = position < changes.size() - 1 && position + 2 < changes.size();
+        return posCheck && isUpdate(changes.get(position + 1), changes.get(position + 2));
+    }
+
+    private ValidityReport validate() {
+        Reasoner reasoner = ReasonerRegistry.getOWLReasoner();
+        reasoner.bindSchema(ontology);
+        InfModel infm = ModelFactory.createInfModel(reasoner, model);
+
+        return infm.validate();
+    }
+
+    private class ChangeListener implements SimpleModelChangeHandler {
+
+        private void logChange(Change ch) {
+
+            // do not log undo changes
+            if (!isUndo(ch) && !isRedo(ch)) {
+
+                // forget all undone changes, if any
+                if (position < changes.size() - 1) {
+                    changes.subList(position, changes.size()).clear();
+                }
+
+                changes.add(ch);
+                position = changes.size() - 1;
+            }
+        }
+
+        private boolean isUndo(Change newCh) {
+            if (changes.size() > 0 && position > -1) {
+                Change oldCh = changes.get(position);
+
+                boolean isEqual = oldCh.getChange().difference(newCh.getChange()).isEmpty();
+                boolean isOpposite = oldCh.isPositive() ^ newCh.isPositive();
+
+                return isEqual && isOpposite;
+            } else {
+                return false;
+            }
+        }
+
+        private boolean isRedo(Change newCh) {
+            if (position < changes.size() - 1) {
+                Change oldCh = changes.get(position + 1);
+
+                boolean isEqual = oldCh.getChange().difference(newCh.getChange()).isEmpty();
+                boolean isOpposite = oldCh.isPositive() == newCh.isPositive();
```

Comment:
This implementation returns false when oldCh.isPositive() and newCh.isPositive() are different and true when they are the same. If this is what is wanted here, the name "isOpposite" is misleading and should be changed.